### DISCUSSION TITLE
Help needed with a SPICE library device

### DISCUSSION THREAD

**Discussion by bnilsson11** [ra3xdh/qucs_s#950] [Q&A] [answered]
Hi,

The "SPICE library device" is usually working well for me, but I have encountered a model (a Schottky diode SB5100) that is not working properly. This model is working in KiCad, but not in qucs-s. I need some help figuring out why.
I am sorry if this is off-topic and hope for the best.

I get the following error in the sim log:

> Warning: Model issue on line:
  .model xsb5100:d1rectifier d is=    1.250000000000000e-05     bv=    1.0 ...
unrecognized parameter (vpk) - ignored
unrecognized parameter (iave) - ignored
unrecognized parameter (vp) - ignored
Error on line:
  d.xsb5100.d1 rvp out xsb5100:d1rectifier 
device already exists, bail out
    Simulation interrupted due to error!

Please see the attached model library.
[sb5100_spice.lib.txt](https://github.com/user-attachments/files/17027572/sb5100_spice.lib.txt)

**Comment by tomhajjar** [comment]
Model works in LTspice but isn't working in ngspice as a model.

See Issue #951

**Comment by ra3xdh** [comment]
Try setting the LTspice compatibility mode globally at the simulation dialog. Probably you have this option enabled somewhere in KiCAD settings.  

![image](https://github.com/user-attachments/assets/b908394f-6848-4f40-bbe5-b0c4869114d1)

**Reply by bnilsson11** [reply]
I tried it and it made no difference.

**Reply by bnilsson11** [reply]
I am now not so sure I really had a problem.
Maybe I defined the device in the wrong way.
I needed two diodes, and I created two SPICE library devices and set them up the same way.
Then I got the error "device exists".
If I instead created ONE diode and set it up using sb5100_spice.lib.txt, it worked.
If I then copy/pasted this working device to create the second one, then both worked.
Is this expected?

**Reply by tomhajjar** [reply]
Can you post your files? @bnilsson11

**Comment by bnilsson11** [comment]
I am very sorry, right now I cannot reproduce it.
It was a part of a fairly complex design, and I tried it in a clean simple version schematic but then there was no problem.
I deleted the SB5100 devices from the complex design and added them again this time also without problem.
From my side you can close this issue, sorry I have taken up your valuable time.

**Comment by ra3xdh** [comment]
I have tried to simulate the schematic using the attached mode and I cannot reproduce the simulation problem. Everything works on my machine using  LTspice compatibility mode. I remember that you are running Qucs-S on Mac. This may some specific issues with Ngspice Mac version. Does this model work using CLI Ngspice mode? The KiCAD uses different Ngspice compiled as shared library. It may have different compile time options. 

Also try the following netlist with Ngspice CLI version:

~~~
* Qucs 24.3.1  /tmp/dio.sch
* Fix the path in the next line
.INCLUDE "/tmp/sb5100_spice.lib"
XX1  _net0 0 SB5100 
V1 _net1 0 DC 1
VPr1 _net1 _net0 DC 0

.control
dc v1 0 0.5 0.0263158
plot i(VPr1)
.endc
~~~


![image](https://github.com/user-attachments/assets/7147b98a-0909-4e01-8cf8-40a02048b3a3)

**Comment by tomhajjar** [comment]
@ra3xdh 

Your test circuit is different than mine. I used a "Red" diode and an .INCLUDE statement pointing at the Spice sub-circuit.  

Shouldn't my configuration work?

![2024-09-18_134521](https://github.com/user-attachments/assets/a563e92c-0dc7-4406-8e70-a08172dc3475)

[Rectifiers_prj.zip](https://github.com/user-attachments/files/17048272/Rectifiers_prj.zip)

**Comment by ra3xdh** [comment]
This will not work. Furthermore, the red diode never worked in a such way. It requires the model (`.MODEL`) name, but the `.SUBCKT` is given in the schematic above.

**Comment by tomhajjar** [comment]
So Red diode is handled differently than Red mosfet/bjt?

![2024-09-18_141818](https://github.com/user-attachments/assets/477be558-f522-472b-8e69-47de6b1420d8)
[BSIM3_prj.zip](https://github.com/user-attachments/files/17048632/BSIM3_prj.zip)

**Comment by ra3xdh** [comment]
All red devices require a `.MODEL` name. The specification of `.SUBCKT` name will not work. It was never accepted. The `N1` in your schematic is the `.MODEL` name defined in `BSIM3v3.lib` file.

**Comment by ra3xdh** [comment]
The `.SUBCKT` is accepted only if you have specified an `Letter=X` letter for MOSFET device or use device named like *X (NMOS)*.

**Comment by tomhajjar** [comment]
I finally see the issue. This is a rare case of a manufacturer making a generic subcircuit instead of a simple .Model. I assume they did this to make models faster. Funny the company only offers Spice models for a few of it's diodes...

Here are two MOSFET amps. One uses MOSFET models and the other MOSFET sub-circuits. I don't see "X" being used anywhere.

[Amp_Current_Dump_MOSFET_prj.zip](https://github.com/user-attachments/files/17048872/Amp_Current_Dump_MOSFET_prj.zip)

**Comment by tomhajjar** [comment, marked as answer]
Fixed the diode project. Created 2 examples.

1) diode as subcircuit
2) diode as Spice lib
 
[Rectifiers_prj.zip](https://github.com/user-attachments/files/17051067/Rectifiers_prj.zip)
